http-propagation-utilities (honey-7):

tomoNeoworks opened the pull request:
# Changes:
- Added high-level utilities to make manual instrumentation of HTTP client/server easier.
- Refactored the Spring starter's ServletFilter and RestTemplate Interceptor to use these.
- Deprecated constants in the Spring starter and moved them to Core where appropriate.
- Configured Mockito to allow mocks on final classes.
- Fixed a versioning issue in the parent POM: the POM was referencing an old version of Core.

# Notes for reviewers
Below are some points that are worth pointing out:
1. AFAIK, when we were creating the Spring HTTP instrumentation, Honeycomb said that the span names generated by this should be low cardinality. As such, the span name generated by the [BeelineRestTemplateInterceptor](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-spring-boot-starter/src/main/java/io/honeycomb/beeline/spring/beans/BeelineRestTemplateInterceptor.java) was a constant and the span names generated by the [BeelineServletFilter](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-spring-boot-starter/src/main/java/io/honeycomb/beeline/spring/beans/BeelineServletFilter.java) were a constant concatenated with the name of the HTTP method being called. For the changes in this PR, I made this low cardinality assumption clear in the documentation. See the constructors of the [HttpClientPropagator](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-core/src/main/java/io/honeycomb/beeline/tracing/propagation/HttpClientPropagator.java) and the [HttpServerPropagator](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-core/src/main/java/io/honeycomb/beeline/tracing/propagation/HttpServerPropagator.java). Please can reviewers verify that the assumption and wording in the documentation is correct.
2. If you compare the documentation at the top of [TraceFieldConstants](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-core/src/main/java/io/honeycomb/beeline/tracing/utils/TraceFieldConstants.java) with the documentation at the top of [MoreTraceFieldConstants](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-spring-boot-starter/src/main/java/io/honeycomb/beeline/spring/utils/MoreTraceFieldConstants.java), it implies that the former contains span fields names that are in the Honeycomb spec and the latter contains span fields name that are not yet. In moving the HTTP instrumentation into `beeline-core` from `beeline-spring-boot-starter` I moved constants from the "non-spec" file to the "spec" file. I am not sure what the current fields in the Honeycomb spec are today, so could you verify that this is desired for each of these constants.
3. I changed the functionality of the BeelineRestTemplateInterceptor slightly. It used to throw an exception if a call to `org.springframework.http.client.ClientHttpResponse#getRawStatusCode`
threw an IOException. I took the decision to say that this should be swallowed at the level of the instrumentation to ensure: (1) we don't have checked exceptions in the 
[HttpClientResponseAdapter](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-core/src/main/java/io/honeycomb/beeline/tracing/propagation/HttpClientResponseAdapter.java) interface for clients to deal with, and (2) we don't break the rule against having the instrumentation throw exceptions up to the non-instrumentation layer. I've added two Spring-specific span fields for noting that this edge case occurred. Please verify that this behaviour is desired.
4. I think having the original [Beeline](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-core/src/main/java/io/honeycomb/beeline/tracing/Beeline.java) class and the new [DefaultBeeline](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-core/src/main/java/io/honeycomb/beeline/DefaultBeeline.java) might be a bit confusing for some clients. The documentation on the Beeline seems to imply it is a place for convenience methods, e.g:
    ```
     * However, access to the low level API is given via the delegates returned by {@link #getTracer()} and
     * {@link #getSpanBuilderFactory()}. The Beeline class is essentially a façade to those delegates,
     * narrowing and bundling lower-level APIs to target the most common use-cases.
    ```
    But the documentation on the DefaultBeeline also suggests it is a place for convenience methods.  However, if we put convenience methods on the DefaultBeeline (as have been added), those methods can't be used if you only have a reference to the Beeline. And if you use the DefaultBeeline, you have to use a bunch of default settings.

    I wanted to add a convenience method (see `TraceStarter#startTrace` inside
[HttpServerPropagator](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-core/src/main/java/io/honeycomb/beeline/tracing/propagation/HttpServerPropagator.java#L126)) but putting it on either the Beeline or DefaultBeeline would create problems at this point, so I put it in a place which did not change the public API. I suggest we discuss what is the best approach to addressing this, and if any change is required we should put it in a different PR.

tredman commented:
FWIW - I'm not really knowledgable about Java patterns. I think any new functionality should go into `Beeline` - we can adapt `DefaultBeeline` to pick up any new functionality if we decide it's worthwhile. `DefaultBeeline` is basically "Beeline lite" - we wanted a way for folks to get started with as few lines of code as possible (see https://github.com/honeycombio/beeline-java/issues/12), while still allowing advanced users a way to configure things as they want. The limitations and default settings in `DefaultBeeline` are an intentional tradeoff to reduce boilerplate, while still letting new users get data in.

asdvalenzuela commented:
Responding to your notes for reviewers:
1. I don’t think that this assumption is correct. Honeycomb does not have issues with high cardinality fields. We also feel that span names should reflect the operation, so if there are a lot of different types of requests, they should be given different names.
2. Yep, it’s fine to move these constants over to the “spec” file. We have slight variations in some of these fields across Beelines, but we do not think it is worth a major version bump to update them at this time.
3. Sounds good.
4. I agree with @tredman here. The `DefaultBeeline` was created to address a real user concern (around reducing the amount of boilerplate) that we were hearing come up repeatedly. If there’s a way to reduce confusion between the two while keeping a simpler implementation (`DefaultBeeline`) and one that is more configurable for advanced users (`Beeline`), we’d be happy to address that. I agree that new functionality can be added to `Beeline` first and then extended to `DefaultBeeline` later.

tomoNeoworks commented:
@asdvalenzuela @tredman On the issues raised:
1. That is good to hear that Honeycomb supports this; I've updated the javadoc to remove the mistaken references to low cardinality.
2. No action required.
3. No action required.
4. I've added the convenience method to just the Beeline for now. The user concern about wanting to reduce boilerplate when starting out with simple manual instrumentation makes complete sense. Core didn't support that as the focus was on using it for Spring auto-instrumentation, but it should. My main worry was that because the two "Beeline" classes are unrelated types we might end up with duplicated convenience methods and confusion over which to use; it looks manageable at the moment though. Perhaps an alternative approach would have been to create a method that returns an instance of `Beeline` with default settings. This `Beeline` instance would not have been a real Singleton, so it depends if that was also a requirement. If the requirement was just global access we could have set it as a "default instance" that was globally accessible (this access pattern is similar to that used to get the HoneyClient in [LibHoney](https://github.com/honeycombio/libhoney-java/blob/master/libhoney/src/main/java/io/honeycomb/libhoney/LibHoney.java)). 

As discussed with @eanakashima we would also like some docs on the new functionality for the Honeycomb site. I will follow-up with those.

tomoNeoworks commented:
@asdvalenzuela Below is some markdown documentation on this feature that you may wish to add in some form to the website. The links would have to be tweaked when the actual feature is merged, because they are linking to the forked repo. I also fixed an error in the javadoc that I noticed when writing this up.

---

# Instrumenting HTTP clients and servers

The Beeline Spring Boot Starter auto-instruments the HTTP server and client inside your Spring Boot application. If you are not using Spring Boot you can still instrument your HTTP server and client code using Beeline Core's HTTP instrumentation utilities.

## Instrumenting HTTP clients

Beeline Core contains high-level utility classes to help you instrument the HTTP client of your choice. Using this pattern will open and close a child span for each HTTP client call and add a standard set of fields to this span based on the data in the HTTP request and response. The class that does the work is [HttpClientPropagator](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-core/src/main/java/io/honeycomb/beeline/tracing/propagation/HttpClientPropagator.java). You pass implementations to it of  [HttpClientRequestAdapter](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-core/src/main/java/io/honeycomb/beeline/tracing/propagation/HttpClientRequestAdapter.java) and [HttpClientResponseAdapter](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-core/src/main/java/io/honeycomb/beeline/tracing/propagation/HttpClientResponseAdapter.java) which [adapt](https://en.wikipedia.org/wiki/Adapter_pattern) the HTTP request and response abstractions used in your HTTP client library. This allows the `HttpClientPropagator` to read and write to the request and response so the client call can be traced. The general pattern is as follows:

```java
public class HttpClientCallInterceptor { 
    // Initialised in constructor
    private final HttpClientPropagator clientPropagator;    

    ///...

    public HttpResponse instrumentedClientCall(final HttpRequest actualHttpRequest) throws Exception {
            // Adapt your actual HTTP request so it can be read by the HttpClientPropagator
            HttpClientRequestAdapter adaptedHttpRequest = new MyHttpClientRequestAdapter(actualHttpRequest);

            // These two variables are so we can reference the adapted response and error (if they exist) when closing the child span using #endPropagation
            HttpClientResponseAdapter adaptedHttpResponse = null;
            Throwable error = null;

            // Call #startPropagation just before you execute the actual HTTP request; this will start the child span and add standard HTTP span fields.
            Span span = clientPropagator.startPropagation(adaptedHttpRequest);

            try {
                // Execute the actual HTTP request
                HttpResponse actualHttpResponse = makeClientCall(actualHttpRequest);

                // Adapt your actual HTTP response so it can be read by the HttpClientPropagator
                adaptedHttpResponse = new MyHttpClientResponseAdapter(actualHttpResponse);

                //Return the actual HTTP response for processing by your code
                return actualHttpResponse;
            } catch (Exception ex) {
                // Assign the error if present
                error = ex;
                throw ex;
            } finally {
                // Call #endPropagation when request has finished. This closes the child span and adds the standard HTTP span fields based on whether the error/response are non-null.
                clientPropagator.endPropagation(adaptedHttpResponse, error, span);
            }
    }
    
    private HttpResponse makeClientCall(HttpRequest httpRequest) {
       	// Make the actual HTTP client call
        /// ...
    }

    class MyHttpClientRequestAdapter implements HttpClientRequestAdapter {

            private final HttpRequest httpRequest;

            public MyHttpClientRequestAdapter(HttpRequest httpRequest) {
                this.httpRequest = httpRequest;
            }

        //implement methods
        ///... 
    }

    class MyHttpClientResponseAdapter implements HttpClientResponseAdapter {

            private final HttpResponse httpResponse;

            public MyHttpClientResponseAdapter(HttpResponse httpResponse) {
                this.httpResponse = httpResponse;
            }

        //implement methods
        ///... 
    }
}
```

It is likely that you would implement this in an interceptor that is added to your HTTP client. An example of this is found in the [instrumentation](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-spring-boot-starter/src/main/java/io/honeycomb/beeline/spring/beans/BeelineRestTemplateInterceptor.java) Beeline adds to Spring's `RestTemplate` interceptor.  This also contains an example of how to adapt an HTTP client request and response.

## Instrumenting HTTP servers

Beeline Core contains high-level utility classes to help you instrument the HTTP server of your choice. Using this pattern will open and close a span for each HTTP request received by the server and add a standard set of fields to this span based on the data in the HTTP request and response. If the HTTP request is being made by a service that is also using Beeline tracing, then the new span will be the child of the HTTP client's request span. The class that does the work is [HttpServerPropagator](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-core/src/main/java/io/honeycomb/beeline/tracing/propagation/HttpServerPropagator.java). You pass implementations to it of  [HttpServerRequestAdapter](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-core/src/main/java/io/honeycomb/beeline/tracing/propagation/HttpServerRequestAdapter.java) and [HttpServerResponseAdapter](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-core/src/main/java/io/honeycomb/beeline/tracing/propagation/HttpServerResponseAdapter.java) which [adapt](https://en.wikipedia.org/wiki/Adapter_pattern) your framework's HTTP request and response abstractions. This allows the `HttpServerPropagator` to read and write to the request/response so the request handling can be traced. The general pattern is as follows:

```java
// Inside your HTTP server handler class, e.g. a Servlet Filter
public class ServerHandler { 

    // Initialised in constructor
    private final HttpServerPropagator serverPropagator; 
    
    ///...

    public HttpResponse instrumentedRequest(final HttpRequest actualHttpRequest) throws Exception {
            // Adapt the actual received HTTP request so it can be read by the HttpServerPropagator
            HttpServerRequestAdapter adaptedHttpRequest = new MyHttpServerRequestAdapter(actualHttpRequest);

            // These two variables are so we can reference the adapted response and error (if they exist) when closing the child span using #endPropagation
            HttpServerResponseAdapter adaptedHttpResponse = null;
            Throwable error = null;

            // Call #startPropagation before you start handling the actual HTTP request; this will start the child span and add standard HTTP span fields.
            Span span = serverPropagator.startPropagation(adaptedHttpRequest);

            try {
                // Handle the HTTP request
                HttpResponse actualHttpResponse = handleRequest(actualHttpRequest);

                // Adapt the actual HTTP response that you will return to the client so it can be read by the HttpServerPropagator
                adaptedHttpResponse = new MyHttpServerResponseAdapter(actualHttpResponse);

                //Return the actual HTTP response to the client
                return actualHttpResponse;
            } catch (Exception ex) {
                // Assign the error if an exception occurred when your code was processing the HTTP request
                error = ex;
                throw ex;
            } finally {
                // Call #endPropagation when the server has finished processing the request. This closes the child span and adds the standard HTTP span fields based on whether the error/response are non-null.
                serverPropagator.endPropagation(adaptedHttpResponse, error, span);
            }
        }

    private HttpResponse handleRequest(HttpRequest actualHttpRequest) {
        ///... handle the actual HTTP request
    }

    class MyHttpServerResponseAdapter implements HttpServerResponseAdapter {

         private final HttpResponse httpResponse;

         public MyHttpServerResponseAdapter(HttpResponse httpResponse) {
             this.httpResponse = httpResponse;
         }
        //implement methods
        ///...     
    }

    class MyHttpServerRequestAdapter implements HttpServerRequestAdapter {

        private final HttpRequest httpRequest;

        public MyHttpServerRequestAdapter(HttpRequest httpRequest) {
            this.httpRequest = httpRequest;
        }
        //implement methods
        ///... 
    }
}
```

Beeline Core already uses this approach to add server instrumentation via a [Java Servlet Filter](https://github.com/wtce/beeline-java/blob/http-trace-propagation-utilities/beeline-spring-boot-starter/src/main/java/io/honeycomb/beeline/spring/beans/BeelineServletFilter.java). This also contains an example of how to adapt an HTTP server request and response.

maplebed commented:
I want to add some context to the "low cardinality" comment as I may have been its source. Span names should be human-useful and reference describe the main reason for the span. It is very useful to group by span name to collect all the spans doing the same thing together. It is common in an incoming HTTP span, for example, to use the name of the route matched to handle the request as the span name. This is a "low cardinality" set in the most general sense - it probably has values in the tens or hundreds or thousands and probably not in the millions.

It is very frustrating when the span name field has arbitrarily random data included it in, such as the pointer address of the function being called or a request ID or something like that; it prohibits any useful grouping using the field. The advice I usually give when someone suggests that the name field should be a combination of values is that concatenating things into fields is a poor choice - include the two things in separate fields and the user can concatenate them at query time (eg route name and HTTP method). 

I hope that helps.

tomoNeoworks commented:
Thanks for the clarification @maplebed.

@asdvalenzuela do you think we should add anything to the javadoc for this? 

If so, if there is a 'best practices' page that talks about span names, then it might be a good approach to link to that. This will avoid us maintaining this advice in the repo. 

Alternatively, we could amend the javadoc as follows: 
```
{@code requestToSpanName} allows you to dynamically name the HTTP server span such that the name reflects the operation. This allows you to usefully group spans together when querying by span name in Honeycomb. 

Best Practices:
- Avoid putting random data in span names as it it prohibits any useful grouping when querying Honeycomb by span name. 
- Remember you can always query Honeycomb by grouping by multiple span fields. Therefore, it is often redundant to produce the span name from just a concatenation span field values.
```
Let us know your favoured approach.

Is there anything additional that this PR needs to get it merged? HONEY-9 (migrating the servlet filter to Core) builds on top of this, so it will be much easier to review (at least within Github) if HONEY-7 gets merged.

asdvalenzuela commented:
Let's merge this now and we can amend the docs later if needed.
